tests/test_integration.py

```diff
@@ -1349,8 +1350,11 @@ def assert_reproducible_build(args):
     # Note that we change the `PYTHONHASHSEED` to ensure that there are no issues resulting
     # from the random seed, such as data structures, as Tox sets this value by default. See
     # https://tox.readthedocs.io/en/latest/example/basic.html#special-handling-of-pythonhashseed.
-    run_pex_command(args + ['-o', pex1], env=make_env(PYTHONHASHSEED=111))
-    run_pex_command(args + ['-o', pex2], env=make_env(PYTHONHASHSEED=22222))
+    def create_pex(path, seed):
+      run_pex_command(args + ['-o', path], env=make_env(PYTHONHASHSEED=seed))
+    create_pex(pex1, seed=111)
+    time.sleep(2)
```

Comment:
This is horrible to have to do. Do you have any understanding why this is needed? I'm just very sensitive to adding sleeps - we've done much too much of that back over in Pants.

Reply:
You beat me to writing my explanation:

I had to add this line to avoid flaky tests. By removing `.pyc`, the exploded Pex becomes equivalent for all but the `bdist` case! 

However, the final assertion on the `.pex` files flakily fails. With an appropriate `time.sleep()`, it consistently fails. I suspect this has to do with file metadata changing, but haven't investigated yet.

Flagging this because I know `time.sleep()` has caused much pain in Pants CI. 2 seconds consistently failed all tests for me on macOS. Should it 3 in case CI is slower than my mac? Or is there a better alternative?

Reply:
Really the only good alternative is to figure out what timing has to do with anything and coment the sleep purpose if it's needed. Thinking about this, it probably is needed and its needed to ensure "now" is different to get different timestamps in bits that have timestamp dependence.

Reply:
OK - feeling better about this - I think that's it. A comment is enough in that case. This is one of the very rare cases sleeping is ok / needed.

Reply:
Discussed offline, but older time.sleep could / can sleep less than requested.

See:
+ documented too little or too much: https://docs.python.org/2/library/time.html#time.sleep
+ fixed to never be too little: https://github.com/python/cpython/commit/79d68f929d8def878766965f513b628023f809b5#diff-0fbe6dc9a4b14052d6815e3ae9dce53eR353

I'd be interested to know if the failures you see using `time.sleep(1)` are only on python < 3.5. If so, yay - full explanation. Also if so, a test utility to use time.sleep + tracking elapsed time to ensure the requested sleep time is met or exceeded would be great. We could keep the sleep both minimal and fully understood.

Reply:
No, this is flaking with Python 3.7 :/ confirmed that 1.1 seconds flakes for me with Py37.

To reproduce, apply this diff:
```diff
diff --git a/pex/bin/pex.py b/pex/bin/pex.py
index 7095833..081e6a2 100755
--- a/pex/bin/pex.py
+++ b/pex/bin/pex.py
@@ -291,6 +291,17 @@ def configure_clp_pex_options(parser):
            'No value (alias for prefer, for backwards compatibility). '
            '[Default: %default]')

+  group.add_option(
+      '--reproducible', '--not-reproducible',
+      dest='reproducible',
+      default=True,
+      action='callback',
+      callback=parse_bool,
+      help='Ensure that the generated Pex file is completely reproducible, i.e. if you were '
+           'to run the same command again, the new PEX would be byte-for-byte equivalent '
+           'to the original. This means that the generated Pex will not have .pyc files '
+           'included, which results in a slight startup performance hit.')
+
   parser.add_option_group(group)


@@ -674,7 +685,7 @@ def main(args=None):
     with TRACER.timed('Building pex'):
       pex_builder = build_pex(reqs, options, resolver_options_builder)

-    pex_builder.freeze()
+    pex_builder.freeze(bytecode_compile=not options.reproducible)
     pex = PEX(pex_builder.path(),
               interpreter=pex_builder.interpreter,
               verify_entry_point=options.validate_ep)
@@ -683,7 +694,7 @@ def main(args=None):
       log('Saving PEX file to %s' % options.pex_name, V=options.verbosity)
       tmp_name = options.pex_name + '~'
       safe_delete(tmp_name)
-      pex_builder.build(tmp_name)
+      pex_builder.build(tmp_name, bytecode_compile=not options.reproducible)
       os.rename(tmp_name, options.pex_name)
     else:
       if not _compatible_with_current_platform(options.platforms):
```

Then this diff to unskip the tests:
```diff
diff --git a/tests/test_integration.py b/tests/test_integration.py
index 24a8ea0..7d08f63 100644
--- a/tests/test_integration.py
+++ b/tests/test_integration.py
@@ -1373,7 +1373,7 @@ def assert_reproducible_build(args):
     assert filecmp.cmp(pex1, pex2, shallow=False)


-@pytest.mark.skip("Acceptance test for landing https://github.com/pantsbuild/pex/issues/716.")
+# @pytest.mark.skip("Acceptance test for landing https://github.com/pantsbuild/pex/issues/716.")
 def test_reproducible_build_no_args():
   assert_reproducible_build([])

@@ -1384,17 +1384,17 @@ def test_reproducible_build_bdist_requirements():
   assert_reproducible_build(['six==1.12.0', 'cryptography==2.6.1'])


-@pytest.mark.skip("Acceptance test for landing https://github.com/pantsbuild/pex/issues/716.")
+# @pytest.mark.skip("Acceptance test for landing https://github.com/pantsbuild/pex/issues/716.")
 def test_reproducible_build_sdist_requirements():
   assert_reproducible_build(['pycparser==2.19', '--no-build'])


-@pytest.mark.skip("Acceptance test for landing https://github.com/pantsbuild/pex/issues/716.")
+# @pytest.mark.skip("Acceptance test for landing https://github.com/pantsbuild/pex/issues/716.")
 def test_reproducible_build_m_flag():
   assert_reproducible_build(['-m', 'pydoc'])


-@pytest.mark.skip("Acceptance test for landing https://github.com/pantsbuild/pex/issues/716.")
+# @pytest.mark.skip("Acceptance test for landing https://github.com/pantsbuild/pex/issues/716.")
 def test_reproducible_build_c_flag():
   setup_py = dedent("""
     from setuptools import setup
@@ -1408,11 +1408,11 @@ def test_reproducible_build_c_flag():
     assert_reproducible_build([project_dir, '-c', 'my_app'])


-@pytest.mark.skip("Acceptance test for landing https://github.com/pantsbuild/pex/issues/716.")
+# @pytest.mark.skip("Acceptance test for landing https://github.com/pantsbuild/pex/issues/716.")
 def test_reproducible_build_python_flag():
   assert_reproducible_build(['--python=python2.7'])


-@pytest.mark.skip("Acceptance test for landing https://github.com/pantsbuild/pex/issues/716.")
+# @pytest.mark.skip("Acceptance test for landing https://github.com/pantsbuild/pex/issues/716.")
 def test_reproducible_build_python_shebang_flag():
   assert_reproducible_build(['--python-shebang=/usr/bin/python'])
```

Then `tox -v -e py37-integration -- -k test_reproducible_build`.

--

Given this stabilization to `time.sleep()` in Py3, I don't think our own custom function should make a difference? I can try if you think it will.

Reply:
It won't make a difference and we don't understand what's going on here. If you can just call that out in a comment near the sleep and perhaps file a linked issue to get to the bottom of it

Reply:
Aha. RIF strikes again. See '4.4.6 date and time fields:' here https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT

I think this is fully solved. We need a two second sleep. Under python < 3.5 we need a sleep helper function.

Reply:
And to make sense of the two seconds, they allocated 5 bits to the second field, so 32 distinct values, so can only represent every 2 seconds: http://www.vsft.com/hal/dostime.htm

"Neat"!